File: src/units/UnitManager.ts
/**
 * Manages all units in the game.
 * Uses object pooling for efficient unit creation/destruction.
 */
import { HexCoordinates } from '../core/HexCoordinates';
import { HexGrid } from '../core/HexGrid';
import { SpatialHash } from '../utils/SpatialHash';
import { ObjectPool } from '../utils/ObjectPool';
import {
  UnitData,
  UnitType,
  UnitStats,
  canTraverseLand,
  canTraverseWater,
} from './UnitTypes';

export class UnitManager {
  private units: Map<number, UnitData> = new Map();
  private spatialHash: SpatialHash<UnitData>;
  private nextId = 1;

  // Object pool for unit data
  private unitPool: ObjectPool<UnitData>;

  // Reference to grid for terrain checks
  private grid: HexGrid;

  constructor(grid: HexGrid) {
    this.grid = grid;
    this.spatialHash = new SpatialHash<UnitData>(2);

    // Create unit pool
    this.unitPool = new ObjectPool<UnitData>(
      () => ({
        id: 0,
        type: UnitType.Infantry,
        q: 0,
        r: 0,
        health: 100,
        maxHealth: 100,
        movement: 2,
        maxMovement: 2,
        attack: 10,
        defense: 8,
        playerId: 0,
        hasMoved: false,
      }),
      (unit) => {
        unit.id = 0;
        unit.q = 0;
        unit.r = 0;
        unit.health = 100;
        unit.movement = 2;
        unit.hasMoved = false;
      }
    );

    // Pre-warm pool
    this.unitPool.prewarm(50);
  }

  /**
   * Create a new unit at the specified hex.
   */
  createUnit(type: UnitType, q: number, r: number, playerId: number): UnitData | null {
    // Check if hex is valid
    const cell = this.grid.getCellAt(q, r);
    if (!cell) {
      return null;
    }

    // Check domain compatibility
    const isWater = cell.elevation < 0;
    if (isWater && !canTraverseWater(type)) {
      return null; // Land unit can't be placed on water
    }
    if (!isWater && !canTraverseLand(type)) {
      return null; // Naval unit can't be placed on land
    }

    // Check if hex is already occupied
    if (this.getUnitAt(q, r)) {
      return null;
    }

    // Get unit from pool and initialize
    const unit = this.unitPool.acquire();
    const stats = UnitStats[type];

    unit.id = this.nextId++;
    unit.type = type;
    unit.q = q;
    unit.r = r;
    unit.health = stats.maxHealth;
    unit.maxHealth = stats.maxHealth;
    unit.movement = stats.maxMovement;
    unit.maxMovement = stats.maxMovement;
    unit.attack = stats.attack;
    unit.defense = stats.defense;
    unit.playerId = playerId;
    unit.hasMoved = false;

    // Add to tracking
    this.units.set(unit.id, unit);
    const worldPos = new HexCoordinates(q, r).toWorldPosition(0);
    this.spatialHash.insert(unit, worldPos.x, worldPos.z);

    return unit;
  }

  /**
   * Remove a unit from the game.
   */
  removeUnit(unitId: number): boolean {
    const unit = this.units.get(unitId);
    if (!unit) return false;

    this.units.delete(unitId);
    this.spatialHash.remove(unit);
    this.unitPool.release(unit);

    return true;
  }

  /**
   * Move a unit to a new hex.
   * @param movementCost - Optional movement cost to deduct (if not provided, doesn't deduct)
   */
  moveUnit(unitId: number, toQ: number, toR: number, movementCost?: number): boolean {
    const unit = this.units.get(unitId);
    if (!unit) return false;

    // Check destination is valid
    const cell = this.grid.getCellAt(toQ, toR);
    if (!cell) return false;

    // Check domain compatibility
    const isWater = cell.elevation < 0;
    if (isWater && !canTraverseWater(unit.type)) {
      return false; // Land unit can't move to water
    }
    if (!isWater && !canTraverseLand(unit.type)) {
      return false; // Naval unit can't move to land
    }

    // Check not occupied
    if (this.getUnitAt(toQ, toR)) return false;

    // Check movement cost if provided
    if (movementCost !== undefined) {
      if (unit.movement < movementCost) {
        return false; // Not enough movement
      }
      unit.movement -= movementCost;
      unit.hasMoved = true;
    }

    // Update spatial hash
    this.spatialHash.remove(unit);

    // Update position
    unit.q = toQ;
    unit.r = toR;

    const newPos = new HexCoordinates(toQ, toR).toWorldPosition(0);
    this.spatialHash.insert(unit, newPos.x, newPos.z);

    return true;
  }

  /**
   * Spend movement points for a unit.
   */
  spendMovement(unitId: number, cost: number): boolean {
    const unit = this.units.get(unitId);
    if (!unit || unit.movement < cost) return false;

    unit.movement -= cost;
    unit.hasMoved = true;
    return true;
  }

  /**
   * Reset movement for all units (called at start of turn).
   */
  resetAllMovement(): void {
    for (const unit of this.units.values()) {
      unit.movement = unit.maxMovement;
      unit.hasMoved = false;
    }
  }

  /**
   * Reset movement for units of a specific player.
   */
  resetPlayerMovement(playerId: number): void {
    for (const unit of this.units.values()) {
      if (unit.playerId === playerId) {
        unit.movement = unit.maxMovement;
        unit.hasMoved = false;
      }
    }
  }

  /**
   * Check if a unit can move (has movement points left).
   */
  canMove(unitId: number): boolean {
    const unit = this.units.get(unitId);
    return unit !== undefined && unit.movement > 0;
  }

  /**
   * Get unit at a specific hex.
   */
  getUnitAt(q: number, r: number): UnitData | null {
    const worldPos = new HexCoordinates(q, r).toWorldPosition(0);
    const units = this.spatialHash.queryRadius(worldPos.x, worldPos.z, 0.5);
    return units.find(u => u.q === q && u.r === r) ?? null;
  }

  /**
   * Get all units within range of a hex.
   */
  getUnitsInRange(q: number, r: number, range: number): UnitData[] {
    const worldPos = new HexCoordinates(q, r).toWorldPosition(0);
    // Approximate world distance from hex range
    const worldRadius = range * 1.8;
    return this.spatialHash.queryRadius(worldPos.x, worldPos.z, worldRadius);
  }

  /**
   * Get unit by ID.
   */
  getUnit(id: number): UnitData | undefined {
    return this.units.get(id);
  }

  /**
   * Get all units.
   */
  getAllUnits(): UnitData[] {
    return Array.from(this.units.values());
  }

  /**
   * Get units for a specific player.
   */
  getPlayerUnits(playerId: number): UnitData[] {
    return this.getAllUnits().filter(u => u.playerId === playerId);
  }

  /**
   * Get unit count.
   */
  get unitCount(): number {
    return this.units.size;
  }

  /**
   * Get pool statistics.
   */
  get poolStats() {
    return this.unitPool.stats;
  }

  /**
   * Clear all units.
   */
  clear(): void {
    for (const unit of this.units.values()) {
      this.unitPool.release(unit);
    }
    this.units.clear();
    this.spatialHash.clear();
  }

  /**
   * Spawn random land units for testing.
   */
  spawnRandomUnits(count: number, playerId: number = 1): number {
    const landCells = this.grid.getAllCells().filter(c => c.elevation >= 0);
    let spawned = 0;

    const landTypes = [UnitType.Infantry, UnitType.Cavalry, UnitType.Archer];

    for (let i = 0; i < count && landCells.length > 0; i++) {
      const idx = Math.floor(Math.random() * landCells.length);
      const cell = landCells[idx];
      const type = landTypes[Math.floor(Math.random() * landTypes.length)];

      if (this.createUnit(type, cell.q, cell.r, playerId)) {
        spawned++;
        landCells.splice(idx, 1); // Remove cell so we don't double-place
      }
    }

    return spawned;
  }

  /**
   * Spawn random naval units for testing.
   */
  spawnRandomNavalUnits(count: number, playerId: number = 1): number {
    const waterCells = this.grid.getAllCells().filter(c => c.elevation < 0);
    let spawned = 0;

    const navalTypes = [UnitType.Galley, UnitType.Warship];

    for (let i = 0; i < count && waterCells.length > 0; i++) {
      const idx = Math.floor(Math.random() * waterCells.length);
      const cell = waterCells[idx];
      const type = navalTypes[Math.floor(Math.random() * navalTypes.length)];

      if (this.createUnit(type, cell.q, cell.r, playerId)) {
        spawned++;
        waterCells.splice(idx, 1);
      }
    }

    return spawned;
  }

  /**
   * Spawn a mix of land and naval units for testing.
   */
  spawnMixedUnits(landCount: number, navalCount: number, playerId: number = 1): { land: number; naval: number } {
    return {
      land: this.spawnRandomUnits(landCount, playerId),
      naval: this.spawnRandomNavalUnits(navalCount, playerId),
    };
  }
}

File: results/csharp_converted/organized/Units/unit_manager.cs
using Godot;
using Godot.Collections;


//# Manages all units in the game.

//# Matches web/src/units/UnitManager.ts
[GlobalClass]
public partial class UnitManager : Godot.RefCounted
{
	public Dictionary Units = new Dictionary{};
	// id -> Unit
	public int NextId = 1;
	public Godot.HexGrid Grid;


	// Spatial indexing for O(1) lookups
	protected Dictionary _HexPositions = new Dictionary{};
	// "q,r" -> Unit (for hex coordinate lookups)
	protected Godot.SpatialHash _SpatialHash;

	// For world coordinate queries (radius, rect)
	// Object pooling for units (reduces GC pressure)
	protected Godot.ObjectPool _UnitPool;

	[Signal]
	public delegate void UnitCreatedEventHandler(Godot.Unit unit);
	[Signal]
	public delegate void UnitRemovedEventHandler(int unit_id);
	[Signal]
	public delegate void UnitMovedEventHandler(Godot.Unit unit, int from_q, int from_r);


	public override void _Init(Godot.HexGrid p_grid)
	{
		Grid = p_grid;

		// Cell size of 2.0 works well for hex grids (slightly larger than hex radius)
		_SpatialHash = SpatialHash.New(2.0);

		// Initialize unit pool (will be set up after _init completes)
		_UnitPool = null;
	}


	//# Set up object pooling (call after construction)
	public void SetupPool()
	{


		// Max pool size
		_UnitPool = ObjectPool.New(, new Callable(this, "_create_unit_for_pool"), new Callable(this, "_reset_unit_for_pool"), 500);
	}


	//# Factory function for unit pool.
	protected Godot.Unit _CreateUnitForPool()
	{
		return Unit.New();
	}


	//# Reset function for unit pool.
	protected void _ResetUnitForPool(Godot.Unit unit)
	{
		unit.ResetForPool();
	}


	//# Create a new unit at the specified hex.
	public Godot.Unit CreateUnit(UnitTypes.Type type, int q, int r, int player_id)
	{

		// Check if hex is valid
		var cell = Grid.GetCell(q, r);
		if(cell == null)
		{
			return null;
		}


		// Check domain compatibility
		var is_water = cell.Elevation < HexMetrics.SEA_LEVEL;
		if(is_water && !UnitTypes.CanTraverseWater(type))
		{
			return null;
		}
		// Land unit can't be placed on water
		if(!is_water && !UnitTypes.CanTraverseLand(type))
		{
			return null;

			// Naval unit can't be placed on land

		}// Check if hex is already occupied
		if(GetUnitAt(q, r) != null)
		{
			return null;
		}


		// Acquire unit from pool (or create directly if pool not set up)
		var unit;
		if(_UnitPool)
		{
			unit = _UnitPool.Acquire();
		}
		else
		{
			unit = Unit.New();
		}
		unit.Id = NextId;
		NextId += 1;
		unit.InitWith(type, q, r, player_id);


		// Add to tracking
		Units[unit.Id] = unit;


		// Add to spatial indexes
		var hex_key = "%d,%d" % new Array{q, r, };
		_HexPositions[hex_key] = unit;
		var world_pos = HexCoordinates.New(q, r).ToWorldPosition(cell.Elevation);
		_SpatialHash.Insert(unit, world_pos.X, world_pos.Z);

		EmitSignal("UnitCreated", unit);

		return unit;
	}


	//# Remove a unit from the game.
	public bool RemoveUnit(int unit_id)
	{
		if(!Units.ContainsKey(unit_id))
		{
			return false;
		}

		var unit = Units[unit_id];


		// Remove from spatial indexes
		var hex_key = "%d,%d" % new Array{unit.Q, unit.R, };
		_HexPositions.Erase(hex_key);
		_SpatialHash.Remove(unit);

		Units.Erase(unit_id);


		// Release unit back to pool for reuse (if pool is set up)
		if(_UnitPool)
		{
			_UnitPool.Release(unit);
		}

		EmitSignal("UnitRemoved", unit_id);
		return true;
	}


	//# Move a unit to a new hex.
	public bool MoveUnit(int unit_id, int to_q, int to_r, int movement_cost =  - 1)
	{
		var unit = Units.Get(unit_id);
		if(unit == null)
		{
			return false;
		}


		// Check destination is valid
		var cell = Grid.GetCell(to_q, to_r);
		if(cell == null)
		{
			return false;
		}


		// Check domain compatibility
		var is_water = cell.Elevation < HexMetrics.SEA_LEVEL;
		if(is_water && !unit.CanTraverseWater())
		{
			return false;
		}
		if(!is_water && !unit.CanTraverseLand())
		{
			return false;
		}


		// Check not occupied
		if(GetUnitAt(to_q, to_r) != null)
		{
			return false;
		}


		// Check movement cost if provided
		if(movement_cost >= 0)
		{
			if(unit.Movement < movement_cost)
			{
				return false;
			}
			unit.SpendMovement(movement_cost);
		}


		// Store old position
		var from_q = unit.Q;
		var from_r = unit.R;


		// Update spatial indexes (remove old, add new)
		var old_hex_key = "%d,%d" % new Array{from_q, from_r, };
		var new_hex_key = "%d,%d" % new Array{to_q, to_r, };
		_HexPositions.Erase(old_hex_key);
		_HexPositions[new_hex_key] = unit;


		// Update position
		unit.Q = to_q;
		unit.R = to_r;


		// Update spatial hash with new world position
		var world_pos = HexCoordinates.New(to_q, to_r).ToWorldPosition(cell.Elevation);
		_SpatialHash.Update(unit, world_pos.X, world_pos.Z);

		EmitSignal("UnitMoved", unit, from_q, from_r);
		return true;
	}


	//# Reset movement for all units (called at start of turn).
	public void ResetAllMovement()
	{
		foreach(Variant unit in Units.Values())
		{
			unit.ResetMovement();
		}
	}


	//# Reset movement for units of a specific player.
	public void ResetPlayerMovement(int player_id)
	{
		foreach(Variant unit in Units.Values())
		{
			if(unit.PlayerId == player_id)
			{
				unit.ResetMovement();
			}
		}
	}


	//# Get unit at a specific hex (O(1) lookup via spatial index).
	public Godot.Unit GetUnitAt(int q, int r)
	{
		var hex_key = "%d,%d" % new Array{q, r, };
		return _HexPositions.Get(hex_key);
	}


	//# Get unit by ID.
	public Godot.Unit GetUnit(int id)
	{
		return Units.Get(id);
	}


	//# Get all units.
	public Array<Unit> GetAllUnits()
	{
		var result = new Array{};
		foreach(Variant unit in Units.Values())
		{
			result.Append(unit);
		}
		return result;
	}


	//# Get units for a specific player.
	public Array<Unit> GetPlayerUnits(int player_id)
	{
		var result = new Array{};
		foreach(Variant unit in Units.Values())
		{
			if(unit.PlayerId == player_id)
			{
				result.Append(unit);
			}
		}
		return result;
	}


	//# Get unit count.
	public int GetUnitCount()
	{
		return Units.Size();
	}


	//# Get counts by domain (land vs naval).
	public Dictionary GetUnitCounts()
	{
		var land = 0;
		var naval = 0;
		foreach(Variant unit in Units.Values())
		{
			if(UnitTypes.GetDomain(unit.Type) == UnitTypes.Domain.NAVAL)
			{
				naval += 1;
			}
			else
			{
				land += 1;
			}
		}
		return new Dictionary{{"land", land},{"naval", naval},};
	}


	//# Clear all units.
	public void Clear()
	{

		// Release all units back to pool (if pool is set up)
		if(_UnitPool)
		{
			foreach(Variant unit in Units.Values())
			{
				_UnitPool.Release(unit);
			}
		}
		Units.Clear();
		_HexPositions.Clear();
		_SpatialHash.Clear();
		NextId = 1;
	}


	//# Get units within a world-coordinate radius (for range attacks, area effects).
	public Array<Unit> GetUnitsInRadius(double world_x, double world_z, double radius)
	{
		var results = new Array{};
		foreach(Variant item in _SpatialHash.QueryRadius(world_x, world_z, radius))
		{
			if(item is Godot.Unit)
			{
				results.Append(item);
			}
		}
		return results;
	}


	//# Get units within a world-coordinate rectangle (for selection box).
	public Array<Unit> GetUnitsInRect(double min_x, double min_z, double max_x, double max_z)
	{
		var results = new Array{};
		foreach(Variant item in _SpatialHash.QueryRect(min_x, min_z, max_x, max_z))
		{
			if(item is Godot.Unit)
			{
				results.Append(item);
			}
		}
		return results;
	}


	//# Get spatial hash statistics for debugging.
	public Dictionary GetSpatialStats()
	{
		return _SpatialHash.GetStats();
	}


	//# Get object pool statistics for debugging.
	public Dictionary GetPoolStats()
	{
		if(_UnitPool)
		{
			return _UnitPool.GetStats();
		}
		return new Dictionary{{"available", 0},{"active", 0},{"created", 0},{"reused", 0},{"peak", 0},{"reuse_rate", 0.0},};
	}


	//# Prewarm unit pool with objects.
	public void PrewarmPool(int count)
	{
		if(_UnitPool)
		{
			_UnitPool.Prewarm(count);
		}
	}


	//# Spawn random land units for testing.
	public int SpawnRandomUnits(int count, int player_id = 1)
	{
		var land_cells = new Array{};
		foreach(HexCell cell in Grid.GetAllCells())
		{
			if(cell.Elevation >= HexMetrics.SEA_LEVEL)
			{
				land_cells.Append(cell);
			}
		}

		var spawned = 0;
		var land_types = UnitTypes.GetLandTypes();

		foreach(int i in GD.Range(count))
		{
			if(land_cells.IsEmpty())
			{
				break;
			}
			var idx = GD.Randi() % land_cells.Size();
			var cell = land_cells[idx];
			var type = land_types[GD.Randi() % land_types.Size()];

			if(CreateUnit(type, cell.Q, cell.R, player_id))
			{
				spawned += 1;
				land_cells.RemoveAt(idx);
			}
		}

		return spawned;
	}


	//# Spawn random naval units for testing.
	public int SpawnRandomNavalUnits(int count, int player_id = 1)
	{
		var water_cells = new Array{};
		foreach(HexCell cell in Grid.GetAllCells())
		{
			if(cell.Elevation < HexMetrics.SEA_LEVEL)
			{
				water_cells.Append(cell);
			}
		}

		var spawned = 0;
		var naval_types = UnitTypes.GetNavalTypes();

		foreach(int i in GD.Range(count))
		{
			if(water_cells.IsEmpty())
			{
				break;
			}
			var idx = GD.Randi() % water_cells.Size();
			var cell = water_cells[idx];
			var type = naval_types[GD.Randi() % naval_types.Size()];

			if(CreateUnit(type, cell.Q, cell.R, player_id))
			{
				spawned += 1;
				water_cells.RemoveAt(idx);
			}
		}

		return spawned;
	}


	//# Spawn a mix of land and naval units for testing.
	public Dictionary SpawnMixedUnits(int land_count, int naval_count, int player_id = 1)
	{
		return new Dictionary{
					{"land", SpawnRandomUnits(land_count, player_id)},
					{"naval", SpawnRandomNavalUnits(naval_count, player_id)},
					};
	}


}